### ISSUE TITLE
Allow `Graph-backed assets` to run in a `single process` on any executor

### ISSUE THREAD

**Issue by ion-elgreco** [dagster-io/dagster#19589]
### What's the use case?

Conceptually Ops and graph backed assets are great, breaking up each step into a modular component that is trackable through the UI, but there is a flaw in the current way Dagster handles them and that is each Op (step) is executed in a new process, therefore an InMemoryIOManager won't work unless you execute your full job with a SingleExecutor, which likely no one is going to do in a complex production setting.

Ideally you should be able to define that a graph backed assets which consists of multiple Ops are executed in a single process so that you can pass objects between those Ops (steps) in memory, persisting those in storage would be silly, costly and slow since we have to send data over the network, write, retrieve data over network and read. Which with very large datasets creeping up to TBs this simply won't work, so you need to be able to pass data in memory between other steps. 

### Ideas of implementation

Allow a highlevel job or definitions parameter that will mark graph backed assets to be spawn in a single process which will run all steps in stead of a process for each step.

### Message from the maintainers

Impacted by this issue? Give it a 👍! We factor engagement into prioritization.

**Comment by ion-elgreco** [comment]
Even though it has been asked before, I still believe this is a quite important thing to add to make Ops useful.. 

References:
- https://github.com/dagster-io/dagster/issues/11101
- https://github.com/dagster-io/dagster/issues/14539

**Comment by ppasquet** [comment]
I agree. Having to pass data through I/O operations is wild to me. Another alternative (alongside running everything in a single process) is to collapse multiple ops into one, which defeats the very purpose of `graph`. I actually wonder who is using InMemoryIOManager in a production environment at the moment.

**Comment by jscheel** [comment]
This is certainly a major issue for using Dagster+, as you are metered on the bandwidth as well.

**Comment by dpeng817** [comment]
This is quite a large lift as it requires making significant changes to dagster's execution model - which is obviously something that needs to be done with quite a bit of care.

It's certainly not ideal, but here's a workaround:

You can define a job which uses the in-memory executor and the in-memory io manager which targets only your graph-backed asset, and then use a sensor to kick off runs of that job based on the upstreams.

```python
import dagster as dg

@dg.asset(io_manager_key="s3_io_manager")
def my_upstream_asset():
    return "blah" # this is sent to s3 or something

upstream_job = dg.define_asset_job("upstream_job", [my_upstream_asset])

@dg.graph_asset(ins={"my_upstream_asset": AssetIn(input_manager_key="s3_io_manager")}, io_manager_key="in_mem_io_manager"
)
def my_downstream():
    # do your in memory computations
    ...

# create a job for your downstream asset which runs the asset using the in-process executor
downstream_job = dg.define_asset_job("downstream_job", [my_downstream], executor=dg.in_process_executor)

# create an asset sensor which will run the job whenever the upstream completes
@dg.asset_sensor(asset_key=my_upstream_asset.key, job=downstream_job)
def my_sensor(context, asset_event):
    yield RunRequest()

defs = Definitions=(
    assets=[my_upstream_asset, my_downstream],
    sensors=[my_sensor],
    executor=..., # whatever you want your usual executor to be, defaults to multiprocess still
    resource_defs={"s3_io_manager": ..., "in_mem_io_manager": dg.mem_io_manager}
)
```

Not pretty, but should get you the behavior you're looking for.
